### PR TITLE
Release/0.0.1

### PR DESCRIPTION
**Pull request by poornima-sivanand**

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    apt-get update -qq && \
+    apt-get install -qq -y --no-upgrade patch && \
+    rm -rf /usr/local/tomcat/webapps/* && \
+    mkdir -p /usr/local/tomcat/conf/Catalina/localhost && \
+    chmod g+w /usr/local/tomcat/conf/Catalina/localhost && \
+    chmod g+w /usr/local/tomcat/webapps && \
+    ( cd /usr/local/tomcat && sha1sum --check /usr/local/tomcat/tomcat.sha1 --status) && \
+    patch --batch --forward --unified  --strip=1 --directory=/usr/local/tomcat --input=/usr/local/tomcat/tomcat.patch && \
+    mkdir -p /usr/local/tomcat/log4j/lib /usr/local/tomcat/log4j/conf && \
+    curl -sSL https://github.com/jfclere/tomcat-kubernetes/raw/master/sample.war -o /usr/local/tomcat/webapps/ROOT.war && \
```
Comment (cvarjao) [inline docker/Dockerfile]: This can be removed. It is just a sample app.

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    curl -sSL https://repo1.maven.org/maven2/org/apache/logging/log4j/log4j-api/$LOG4J_VERSION/log4j-api-$LOG4J_VERSION.jar -o /usr/local/tomcat/log4j/lib/log4j-api-$LOG4J_VERSION.jar && \
+    curl -sSL https://repo1.maven.org/maven2/org/apache/logging/log4j/log4j-core/$LOG4J_VERSION/log4j-core-$LOG4J_VERSION.jar -o /usr/local/tomcat/log4j/lib/log4j-core-$LOG4J_VERSION.jar && \
+    curl -sSL https://repo1.maven.org/maven2/org/apache/logging/log4j/log4j-appserver/$LOG4J_VERSION/log4j-appserver-$LOG4J_VERSION.jar -o /usr/local/tomcat/log4j/lib/log4j-appserver-$LOG4J_VERSION.jar && \
+    find /usr/local/tomcat/webapps/ -maxdepth 1 -name '*.war' | xargs basename -s .war | xargs -t -I {} mkdir '/usr/local/tomcat/webapps/{}' && \
```
Comment (cvarjao) [inline docker/Dockerfile]: remove this line. webapps folder should be empty.

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    find /usr/local/tomcat/webapps/ -maxdepth 1 -name '*.war' | xargs basename -s .war | xargs -t -I {} unzip -n -qq '/usr/local/tomcat/webapps/{}.war' -d '/usr/local/tomcat/webapps/{}' && \
```
Comment (cvarjao) [inline docker/Dockerfile]: remove this line. webapps folder should be empty.

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    find /usr/local/tomcat/webapps/ -maxdepth 1 -name '*.war' | xargs -t -I {} rm '{}' && \
```
Comment (cvarjao) [inline docker/Dockerfile]: remove this line. webapps folder should be empty.

File: docker/contrib/original/conf/server.xml [hunk 5]
```diff
@@ -0,0 +1,167 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cvarjao) [inline docker/contrib/original/conf/server.xml]: Remove this file

File: docker/contrib/patched/conf/server.xml [hunk 6]
```diff
@@ -0,0 +1,182 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cvarjao) [inline docker/contrib/patched/conf/server.xml]: Remove this file

File: docker/contrib/original/conf/web.xml [hunk 7]
```diff
@@ -0,0 +1,4737 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cvarjao) [inline docker/contrib/original/conf/web.xml]: Remove this file

File: docker/contrib/patched/log4j/conf/log4j2-tomcat.xml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<Configuration status="error">
```
Comment (cvarjao) [inline docker/contrib/patched/log4j/conf/log4j2-tomcat.xml]: Remove this file

File: docker/contrib/patched/conf/web.xml [hunk 9]
```diff
@@ -0,0 +1,4737 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cvarjao) [inline docker/contrib/patched/conf/web.xml]: Remove this file

File: openshift/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+objects:
+- apiVersion: v1
+  kind: ServiceAccount
+  metadata:
+    creationTimestamp: null
+    name: tomcat
+- kind: Role
+  apiVersion: rbac.authorization.k8s.io/v1
+  metadata:
+    name: tomcat
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Introduce `NAME` and `SUFFIX` parameters and use them in .metadata.name and related fields.

File: openshift/build.yaml [hunk 11]
```diff
@@ -0,0 +1,74 @@
+apiVersion: v1
+kind: Template
+metadata: {}
+objects:
+- apiVersion: image.openshift.io/v1
+  kind: ImageStream
+  metadata:
+    name: tomcat9
```
Comment (cvarjao) [inline openshift/build.yaml:8]: Introduce `NAME` and `SUFFIX` parameters and use them in .metadata.name and related fields.

**Review by cvarjao — COMMENTED** [review]
Missing some files from the bin folder:
https://github.com/cvarjao-o/showcase-java-web-app/tree/release/0.0.1/openshift/images/tomcat/contrib/patched/bin

don't forget to run "create-patch.sh" to keep the patch file updated which is the only file actually committed.

**Comment by poornima-sivanand** [comment]
@cvarjao : Can you please check if it is ok?

File: .gitignore [hunk 12]
```diff
@@ -21,3 +21,11 @@
 
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+*.xml
```
Comment (cvarjao) [inline .gitignore]: This ignore is too broad. Ignoring all .xml files may leave files behind without being committed. Please be more specific.

File: .gitignore [hunk 13]
```diff
@@ -21,3 +21,11 @@
...
+*.sh
```
Comment (cvarjao) [inline .gitignore]: This ignore is too broad. Ignoring all .sh files may leave files behind without being committed. Please be more specific.

File: openshift/deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+  rules:
+  - apiGroups: [""] # "" indicates the core API group
+    resources: ["pods"]
+    verbs: ["get", "watch", "list"]
+- apiVersion: rbac.authorization.k8s.io/v1
+  kind: RoleBinding
+  metadata:
+    creationTimestamp: null
+    name: tomcat
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to use `${NAME}${SUFFIX}`

File: openshift/deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,160 @@
+apiVersion: v1
+kind: Template
+metadata: {}
+objects:
+- apiVersion: v1
+  kind: ServiceAccount
+  metadata:
+    creationTimestamp: null
+    name: tomcat
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to use `${NAME}${SUFFIX}`

File: openshift/deployment.yaml [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+- kind: Role
+  apiVersion: rbac.authorization.k8s.io/v1
+  metadata:
+    name: tomcat
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to use `${NAME}${SUFFIX}`

File: openshift/deployment.yaml [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+      openshift.io/generated-by: OpenShiftNewApp
+    creationTimestamp: null
+    labels:
+      app: ${NAME}
+    name: ${NAME}${SUFFIX}
+  spec:
+    replicas: 1
+    selector:
+      app: ha-tomcat9
+      deploymentconfig: ha-tomcat9
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to refer to the dc.metadata.name which includes`${NAME}${SUFFIX}`

File: openshift/deployment.yaml [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    strategy:
+      resources: {}
+    template:
+      metadata:
+        annotations:
+          openshift.io/generated-by: OpenShiftNewApp
+        creationTimestamp: null
+        labels:
+          app: ha-tomcat9
+          deploymentconfig: ha-tomcat9
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to refer to the dc.metadata.name which includes${NAME}${SUFFIX}

File: openshift/deployment.yaml [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        containers:
+        - name: tomcat
+          image: ' '
+          env:
+            - name: OPENSHIFT_KUBE_PING_NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: OPENSHIFT_KUBE_PING_LABELS
+              value: "deploymentconfig=ha-tomcat9"
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to refer to the dc.metadata.name which includes${NAME}${SUFFIX}

File: openshift/deployment.yaml [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+    test: false
+    triggers:
+    - type: ConfigChange
+    - imageChangeParams:
+        automatic: true
+        containerNames:
+        - tomcat
+        from:
+          kind: ImageStreamTag
+          name: ha-tomcat9:latest
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to use `${NAME}` parameter and some parameter for the ImageTag

File: openshift/deployment.yaml [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+    updatedReplicas: 0
+- apiVersion: v1
+  kind: Service
+  metadata:
+    annotations:
+      openshift.io/generated-by: OpenShiftNewApp
+    creationTimestamp: null
+    labels:
+      app: ha-tomcat9
+    name: ha-tomcat9
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to include `${NAME}` and/or `${SUFFIX}` parameters

File: openshift/deployment.yaml [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    - name: 8443-tcp
+      port: 8443
+      protocol: TCP
+      targetPort: 8443
+    - name: 8778-tcp
+      port: 8778
+      protocol: TCP
+      targetPort: 8778
+    selector:
+      deploymentconfig: ha-tomcat9
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to include `${NAME}` and/or `${SUFFIX}` parameters

File: openshift/deployment.yaml [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+    loadBalancer: {}
+- apiVersion: route.openshift.io/v1
+  kind: Route
+  metadata:
+    annotations:
+      haproxy.router.openshift.io/balance: roundrobin
+      haproxy.router.openshift.io/disable_cookies: 'true'
+    labels:
+      app: ha-tomcat9
+    name: ha-tomcat9
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to include `${NAME}` and/or `${SUFFIX}` parameters

File: openshift/deployment.yaml [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+  spec:
+    port:
+      targetPort: 8080-tcp
+    to:
+      kind: Service
+      name: ha-tomcat9
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Needs to include `${NAME}` and/or `${SUFFIX}` parameters

**Review by cvarjao — DISMISSED** [review]

File: openshift/deployment.yaml [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+        annotations:
+          openshift.io/generated-by: OpenShiftNewApp
+        creationTimestamp: null
+        labels:
+          app: ${NAME}
+          deploymentconfig: ${NAME}${SUFFIX}
+      spec:
+        serviceAccountName: "${NAME}${SUFFIX}"
+        containers:
+        - name: ${NAME}${SUFFIX}
```
Comment (cvarjao) [inline openshift/deployment.yaml]: Container names don't need to be unique within the namespace, just within the pod. You can just use "tomcat" as container name. Don't forget to update `ConfigChange` trigger a few lines below.

**Comment by cvarjao** [comment]
@poornima-sivanand , missing `log4j/conf/log4j2-tomcat.xml` file. 

Reference:
- https://github.com/cvarjao-o/showcase-java-web-app/blob/release/0.0.1/openshift/images/tomcat/contrib/patched/log4j/conf/log4j2-tomcat.xml
- https://github.com/cvarjao-o/showcase-java-web-app/blob/release/0.0.1/openshift/images/tomcat/contrib/tomcat.patch#L322

**Comment by poornima-sivanand** [comment]
@cvarjao : Resolved

**Review by mbystedt — APPROVED** [review]

**Pull request merged**